dotnet-desktop-guide/net/wpf/events/how-to-create-a-custom-routed-event.md

```diff
@@ -0,0 +1,82 @@
+---
+<!-- The acrolinx score was 93 on 02/02/2021-->
+
+# How to create a custom routed event (WPF .NET)
+
+Windows Presentation Foundation (WPF) application developers and component authors can create custom routed events to extend the functionality of common language runtime (CLR) events. For information on routed event capabilities, see [Why use routed events](/dotnet/desktop/wpf/advanced/routed-events-overview?view=netframeworkdesktop-4.8&preserve-view=true#why-use-routed-events). This article covers the basics of creating a custom routed event.
+
+[!INCLUDE [desktop guide under construction](../../includes/desktop-guide-preview-note.md)]
+
+## Routed event steps
+
+The basic steps to create a routed event are:
+
+1. Register a <xref:System.Windows.RoutedEvent> using the <xref:System.Windows.EventManager.RegisterRoutedEvent%2A> method.
+
+1. The registration call returns a `RoutedEvent` instance, known as a routed event identifier, which holds the registered event name, [routing strategy](/dotnet/desktop/wpf/advanced/routed-events-overview?view=netframeworkdesktop-4.8&preserve-view=true#routing-strategies), and other event details. Assign the identifier to a static readonly field. By convention:
+    - The identifier for a routed event with a [bubbling](<xref:System.Windows.RoutingStrategy.Bubble>) strategy is named `<event name>Event`. For example, if the event name is `Tap` then the identifier should be named `TapEvent`.
+    - The identifier for a routed event with a [tunneling](<xref:System.Windows.RoutingStrategy.Tunnel>) strategy is named `Preview<event name>Event`. For example, if the event name is `Tap` then the identifier should be named `PreviewTapEvent`.
+
+1. Define CLR [add](<xref:System.Windows.UIElement.AddHandler%2A>) and [remove](<xref:System.Windows.UIElement.RemoveHandler%2A>) event accessors. Without CLR event accessors, you'll only be able to add or remove event handlers through direct calls to the <xref:System.Windows.UIElement.AddHandler%2A?displayProperty=nameWithType> and <xref:System.Windows.UIElement.RemoveHandler%2A?displayProperty=nameWithType> methods. With CLR event accessors, you gain these event handler assignment mechanisms:
+
+    - For Extensible Application Markup Language (XAML), you can use attribute syntax to add event handlers.
+    - For C#, you can use the `+=` and `-=` operators to add or remove event handlers.
+    - For VB, you can use the [AddHandler](/dotnet/visual-basic/language-reference/statements/addhandler-statement) and [RemoveHandler](/dotnet/visual-basic/language-reference/statements/removehandler-statement) statements to add or remove event handlers.
+
+1. Add custom logic for triggering your routed event. For example, your logic might trigger the event based on user-input and application state.
+
+## Example
+
+The following example implements the `CustomButton` class in a custom control library. The `CustomButton` class, which derives from <xref:System.Windows.Controls.Button>:
```

Comment:
Is there a reason why this is a custom control library? It's simpler to have it all in a single project.

Reply:
The framework version of this article says:

`Note also that this example basically implements an entire subclass of Button; that subclass is built as a separate assembly and then instantiated as a custom class on a separate XAML page. This is to illustrate the concept that subclassed controls can be inserted into trees composed of other controls, and that in this situation, custom events on these controls have the very same event routing capabilities as any native XAML element does. `

Someone implementing a custom control that way might benefit, although it adds to the complexity of the example as you mentioned. What's your preference?